Traceback (most recent call last):
  File "/opt/miniconda3/lib/python3.12/site-packages/jupyter_cache/executors/utils.py", line 58, in single_nb_execution
    executenb(
  File "/opt/miniconda3/lib/python3.12/site-packages/nbclient/client.py", line 1314, in execute
    return NotebookClient(nb=nb, resources=resources, km=km, **kwargs).execute()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/lib/python3.12/site-packages/jupyter_core/utils/__init__.py", line 165, in wrapped
    return loop.run_until_complete(inner)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/lib/python3.12/asyncio/base_events.py", line 691, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/opt/miniconda3/lib/python3.12/site-packages/nbclient/client.py", line 709, in async_execute
    await self.async_execute_cell(
  File "/opt/miniconda3/lib/python3.12/site-packages/nbclient/client.py", line 1062, in async_execute_cell
    await self._check_raise_for_error(cell, cell_index, exec_reply)
  File "/opt/miniconda3/lib/python3.12/site-packages/nbclient/client.py", line 918, in _check_raise_for_error
    raise CellExecutionError.from_cell_and_msg(cell, exec_reply_content)
nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
------------------
import numpy as np
from PipeEnv import PipeHeaterSafeEnvGymnasium
from MPCCon import MPCController
from params import *
import matplotlib.pyplot as plt

params["N"] = 30
env = PipeHeaterSafeEnvGymnasium(params, Tinlet=15.0, Tset= 85.0, T_ext=10.0, dt=1.0, t_final=3600)

mpc_dt = 10.0
mpc = MPCController(params, dt=mpc_dt, Hp=80, Q=[0.0, 1e3], R=1e-5, Rd=1.0, T_set=[92.0, 85.0], T_safety=95.0, ramped = True)

obs, _ = env.reset()
------------------


[0;31m---------------------------------------------------------------------------[0m
[0;31mModuleNotFoundError[0m                       Traceback (most recent call last)
Cell [0;32mIn[1], line 3[0m
[1;32m      1[0m [38;5;28;01mimport[39;00m [38;5;21;01mnumpy[39;00m [38;5;28;01mas[39;00m [38;5;21;01mnp[39;00m
[1;32m      2[0m [38;5;28;01mfrom[39;00m [38;5;21;01mPipeEnv[39;00m [38;5;28;01mimport[39;00m PipeHeaterSafeEnvGymnasium
[0;32m----> 3[0m [38;5;28;01mfrom[39;00m [38;5;21;01mMPCCon[39;00m [38;5;28;01mimport[39;00m MPCController
[1;32m      4[0m [38;5;28;01mfrom[39;00m [38;5;21;01mparams[39;00m [38;5;28;01mimport[39;00m [38;5;241m*[39m
[1;32m      5[0m [38;5;28;01mimport[39;00m [38;5;21;01mmatplotlib[39;00m[38;5;21;01m.[39;00m[38;5;21;01mpyplot[39;00m [38;5;28;01mas[39;00m [38;5;21;01mplt[39;00m

File [0;32m~/PycharmProjects/localBook/mybook/heat_pipe_codes/MPCCon.py:3[0m
[1;32m      1[0m [38;5;66;03m# mpc_controller.py (only the MPCController class shown, replace original)[39;00m
[1;32m      2[0m [38;5;28;01mimport[39;00m [38;5;21;01mnumpy[39;00m [38;5;28;01mas[39;00m [38;5;21;01mnp[39;00m
[0;32m----> 3[0m [38;5;28;01mimport[39;00m [38;5;21;01mcvxpy[39;00m [38;5;28;01mas[39;00m [38;5;21;01mcp[39;00m
[1;32m      6[0m [38;5;28;01mclass[39;00m [38;5;21;01mMPCController[39;00m:
[1;32m      7[0m [38;5;250m    [39m[38;5;124;03m"""[39;00m
[1;32m      8[0m [38;5;124;03m    Linear MPC for PipeHeaterSafeEnvGymnasium.[39;00m
[1;32m      9[0m 
[0;32m   (...)[0m
[1;32m     15[0m [38;5;124;03m    - u is heater power in kW (same units as env action); internally env multiplies by 1e3.[39;00m
[1;32m     16[0m [38;5;124;03m    """[39;00m

[0;31mModuleNotFoundError[0m: No module named 'cvxpy'

